Enable folding sections when run under Travis

cognifloyd opened the pull request:
Add output when molecule is running in travis-ci to make the logs easier to jump through.
Travis log folding is explained [here](https://github.com/travis-ci/travis-ci/issues/2285).

#### PR Type
- Feature Pull Request

cognifloyd commented:
Test failures are unrelated to this change.

ssbarnea reviewed (changes requested):
Nice feature, sadly is not yet a generic terminal feature. What worries me is that in case of exceptions, the end section will never be printed, likely causing other issues, especially in cases where multiple commands are executed and some would fail.

Can we maybe catch the exceptions print them and the done bit end and raise them back? 

I am also inclined to opt-in for an implementation using a decorator if possible? Using a decorator should make it much easier to wrap the start/end of it, don't you think?

greg-hellings commented:
This bit concerns me, "While not officially documented, I think you can rely on this behavior in the foreseeable future." But that was written in 2014 and appears to still be valid. So it doesn't concern me that much.

Does it make sense to provide an output mechanism like this specifically just for one CI provider? Do others have similar functionality that we ought to incorporate?

I have no issues with the structure of the code, unless we think we're going to be asked or expected to expand this functionality to support a bunch of other providers.

cognifloyd commented:
> Can we maybe catch the exceptions print them and the done bit end and raise them back?
>
> I am also inclined to opt-in for an implementation using a decorator if possible? Using a decorator should make it much easier to wrap the start/end of it, don't you think?

I don't like requiring all of the commands to add an extra print function at the end. Doing this in a decorator would increase the complexity of the Base command subclasses even more.

What if I modified `execute_scenario` instead, wrapping the call to `execute_subcommand` in a `try`/`finally` block to print the end of the fold line. Or it could be a func wrapper around `execute_subcommand`. I think I'll try a functools wrapper around `execute_subcommand`.

cognifloyd commented:
@ssbarnea Using a decorator seems to be much simpler. Thanks for the idea. What do you think?

@greg-hellings if we do get requests to add something like this for other CI runners (I haven't seen a feature like this anywhere else), then the new wrapper isolates the CI-specific code more so that it's easy to tell what it is doing.

cognifloyd commented:
To test this, I just started upgrading a molecule scenario to use molecule 3.1(ish) - my upgrade is not complete, but it shows how the folding looks:
https://travis-ci.com/github/cognifloyd/community.mongodb/builds/187516426

Sadly, it is not very informative at the folds. I need to play with that to see if there's a way to describe the fold when it is collapsed.

cognifloyd commented:
Ooh. That looks better.
https://travis-ci.com/github/cognifloyd/community.mongodb/builds/187516426

I used this chain of links to learn how to do that:
https://github.com/travis-ci/docs-travis-ci-com/issues/949
https://github.com/travis-ci/travis-rubies/blob/9f7962a881c55d32da7c76baefc58b89e3941d91/build.sh#L38-L44

cognifloyd commented:
OK. So, it's bike-shedding time:

### Text

1. `Molecule Scenario: 'default' Action: 'converge'`
2. `Molecule - Scenario: 'default' - Action: 'converge'`
3. `Molecule | Scenario: 'default' | Action: 'converge'`
4. `Molecule --> Scenario: 'default' --> Action: 'converge'`
5. `--> Scenario: 'default' --> Action: 'converge'`

### Color

Yellow bold (`33;1m`): _Current color_
![Selection_089](https://user-images.githubusercontent.com/1558590/94762979-b9988b00-036e-11eb-8cfe-b8a0ce7ab76f.png)

Yellow normal (`33m`):
![Selection_088](https://user-images.githubusercontent.com/1558590/94762815-4db62280-036e-11eb-8d19-71833688b7bc.png)

White bold (`37;1m`):
![Selection_084](https://user-images.githubusercontent.com/1558590/94761916-eac38c00-036b-11eb-97af-efbdcce21080.png)

White normal (`37m`):
![Selection_085](https://user-images.githubusercontent.com/1558590/94762067-50177d00-036c-11eb-850e-6698b597c323.png)

Cyan bold (`36;1m`):
![Selection_086](https://user-images.githubusercontent.com/1558590/94762419-3f1b3b80-036d-11eb-8dcd-f1179d2ed12b.png)

Cyan normal (`36m`): _oops. I was hovering when I took this screenshot_
![Selection_087](https://user-images.githubusercontent.com/1558590/94762661-f7e17a80-036d-11eb-9c6a-98f6c8a79da3.png)

cognifloyd commented:
My top 3:

1. Cyan bold
2. White normal
3. Yellow bold

cognifloyd commented on molecule/command/base.py:
```diff
@@ -122,6 +123,31 @@ def execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args=()
+    @functools.wraps(func)
+    def travis_wrapper(config, subcommand):
+        print(
+            (
+                "travis_fold:start:{0}.{1}"
+                "\033[33;1mMolecule Scenario: '{0}' Action: '{1}'\033[0m"
+            ).format(
+                config.scenario.name,
+                subcommand,
+            )
```
```suggestion
            (
                "travis_fold:start:{0}.{1}"
                "{2}Molecule Scenario: '{0}' Action: '{1}'{3}"
            ).format(
                config.scenario.name,
                subcommand,
                colorama.Style.Bold + colorama.Fore.CYAN,
                colorama.Style.RESET_ALL,
            )
```

We might also be able to use `colorama` like in logger.py (would need to test to make sure the escape sequence works in this context (colorama uses `\x1b` fwict but the example I found uses `\033`).

ssbarnea commented:
I would not spend much time investing in colorama as I am already working on a big change that drops it and switching to https://github.com/willmcgugan/rich

cognifloyd commented:
Yes, the colorama escape sequence works just fine for Travis. Feel free to replace it with rich when you make the switch.

I went with Bold Cyan with the text `Molecule --> Scenario: 'default' --> Action: 'converge'`:
![Selection_090](https://user-images.githubusercontent.com/1558590/94834366-e7fd8100-03d5-11eb-8a37-88c11ddb2ce7.png)

So, unless anyone else wants to join in the bike-shedding, I think this is ready to merge.

cognifloyd commented:
I squashed and rebased.

Plus I've also prepared a [2.22 backport](https://github.com/cognifloyd/molecule/tree/travis-folding-2.22). Will there be a 2.23 release at some point?

ssbarnea reviewed (changes requested):
That change is far from ready, being broken and not tested:
```
$ molecule list                                                                                                                 [18:36:45]
--> Found config file /Users/ssbarnea/.config/molecule/config.yml
travis_fold:start:default.listMolecule --> Scenario: 'default' --> Action: 'list'
travis_fold:end:default.list
Traceback (most recent call last):
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/bin/molecule", line 33, in <module>
    sys.exit(load_entry_point('molecule', 'console_scripts', 'molecule')())
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/ssbarnea/c/a/molecule/molecule/command/list.py", line 111, in list
    statuses.extend(base.execute_subcommand(scenario.config, subcommand))
TypeError: 'NoneType' object is not iterable
$ molecule list                                                                                                                 [18:36:45]
--> Found config file /Users/ssbarnea/.config/molecule/config.yml
travis_fold:start:default.listMolecule --> Scenario: 'default' --> Action: 'list'
travis_fold:end:default.list
Traceback (most recent call last):
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/bin/molecule", line 33, in <module>
    sys.exit(load_entry_point('molecule', 'console_scripts', 'molecule')())
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 829, in __call__
    return self.main(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 782, in main
    rv = self.invoke(ctx)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1259, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 1066, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/core.py", line 610, in invoke
    return callback(*args, **kwargs)
  File "/Users/ssbarnea/.pyenv/versions/3.8.2/lib/python3.8/site-packages/click/decorators.py", line 21, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/ssbarnea/c/a/molecule/molecule/command/list.py", line 111, in list
    statuses.extend(base.execute_subcommand(scenario.config, subcommand))
TypeError: 'NoneType' object is not iterable
```

cognifloyd commented on molecule/command/base.py:
```diff
@@ -122,6 +124,33 @@ def execute_cmdline_scenarios(scenario_name, args, command_args, ansible_args=()
+            )
+        )
+        try:
+            return func(config, subcommand)
```
I missed this return here. Added. Please try again.

cognifloyd replied:
For background, I discovered that I missed this in the 2.22 branch, but didn't check to see if it was also required for 3.x. Oops.
See: https://github.com/cognifloyd/molecule/commit/6ddc9b2717c43a89bacf24ef7d4a04b498d0fe90

cognifloyd commented:
Should I add a functional test that runs on travis to test this code? Or add a functional test that triggers this by setting the TRAVIS and CI vars?

cognifloyd commented:
I parametrized the unit test for `execute_subcommand` so that it tests as if running on Travis as well.
Do you want any functional or other tests for this? I suppose we could run this and check that the prints happen, but I'm not sure that has much value.

ssbarnea commented:
Please rebase it, I moved molecule/ under lib/molecule. Sorry for the inconvenience.

ssbarnea commented:
Closing due to lack of updates. Ping me if you want to reopen it.

cognifloyd commented:
revisiting now. Please reopen.

cognifloyd commented:
Here's what it looks like with the latest `rich` formatted output instead of straight colorama.
![image](https://user-images.githubusercontent.com/1558590/99199169-e6f49900-2762-11eb-9039-4ff12e12f5c7.png)

cognifloyd commented:
Re-submitted PR as #2967
